<a href="https://colab.research.google.com/github/anasmita3/SURGE/blob/main/SURGE_11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from huggingface_hub import list_repo_files

files = list_repo_files(
    "ai4bharat/IndicCorpV2",
    repo_type="dataset"
)

print(len(files))
print(files[:20])

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:124: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


28
['.gitattributes', 'README.md', 'data/as.txt', 'data/bd.txt', 'data/bn.txt', 'data/dg.txt', 'data/en.txt', 'data/gom.txt', 'data/gu.txt', 'data/hi-1.txt', 'data/hi-2.txt', 'data/hi-3.txt', 'data/kha.txt', 'data/kn.txt', 'data/ks.txt', 'data/mai.txt', 'data/ml.txt', 'data/mni.txt', 'data/mr.txt', 'data/ne.txt']


In [ ]:
from datasets import load_dataset

dataset = load_dataset(
    "ai4bharat/IndicCorpV2",
    data_files={"train": "data/bn.txt"}
)

print(dataset)

README.md:   0%|          | 0.00/4.01k [00:00<?, ?B/s]

data/bn.txt:   0%|          | 0.00/16.0G [00:00<?, ?B/s]

Generating train split: 0 examples [00:00, ? examples/s]

Loading dataset shards:   0%|          | 0/33 [00:00<?, ?it/s]

DatasetDict({
    train: Dataset({
        features: ['text'],
        num_rows: 41004792
    })
})


In [ ]:
print(dataset["train"][0])

{'text': 'নয়াদিল্লি: আজ ভ্যালেন্টাইনস ডে। এমন একটি দিনে প্রিয় মানুষকে হৃদয় উজাড় করা ভালোবাসার বার্তা দিতে পৌঁছে দিতে উন্মুখ অনেকেই। এমন একটি দিনে ভারতীয় দলের প্রাক্তন ক্রিকেটার সচিন তেন্ডুলকর জানালেন তাঁর প্রথম ভালোবাসার কথা। একটি ভিডিও পোস্ট করেছেন ভারতীয় দলের প্রাক্তন ব্যাটিং স্তম্ভ। ভিডিওতে আন্তর্জাতিক ক্রিকেট ১০০ শতরানের মালিককে নেটে অনুশীলন করতে দেখা যাচ্ছে। ওই ভিডিও শেয়ার করে সচিন লিখেছেন, আমার প্রথম ভালোবাসা।'}


In [ ]:
print(dataset)
print(len(dataset["train"]))

DatasetDict({
    train: Dataset({
        features: ['text'],
        num_rows: 41004792
    })
})
41004792


In [ ]:
import time

start = time.time()

for i, row in enumerate(dataset["train"]):
    if i % 100000 == 0:
        print(i, "rows processed")

    if i == 500000:
        break

print("Time:", time.time() - start)

0 rows processed
100000 rows processed
200000 rows processed
300000 rows processed
400000 rows processed
500000 rows processed
Time: 10.057662010192871


In [ ]:
import unicodedata
import re

MAX_DOCS = 500000

with open("bn_news.txt", "w", encoding="utf-8") as f:

    for i, row in enumerate(dataset["train"]):

        text = unicodedata.normalize("NFC", row["text"])
        text = re.sub(r"\s+", " ", text).strip()

        if text:
            f.write(text + "\n")

        if i + 1 >= MAX_DOCS:
            break

print("Saved", MAX_DOCS, "documents")

Saved 500000 documents


In [ ]:
word_count = 0
char_count = 0

with open("bn_news.txt", "r", encoding="utf-8") as f:
    for line in f:
        word_count += len(line.split())
        char_count += len(line)

print("Words:", word_count)
print("Characters:", char_count)
print("Approx MB:", char_count / 1024 / 1024)

Words: 11305065
Characters: 74566089
Approx MB: 71.1117639541626


In [ ]:
import gc

try:
    del dataset
except:
    pass

gc.collect()

36

In [ ]:
!pip -q install aksharamukha

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 289.9/289.9 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.1/183.1 kB 12.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 44.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 94.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 56.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 531.9/531.9 kB 29.2 MB/s eta 0:00:00


In [ ]:
!ls -lh

total 190M
-rw-r--r-- 1 root root 190M Jun 30 04:20 bn_news.txt
drwxr-xr-x 1 root root 4.0K Jun  4 13:32 sample_data


In [ ]:
count = 0

with open("bn_news.txt", "r", encoding="utf-8") as f:
    for _ in f:
        count += 1

print("Documents:", count)

Documents: 250000


In [ ]:
import random

random.seed(42)

train_count = 0
test_count = 0

with open("bn_news.txt", "r", encoding="utf-8") as inp, \
     open("train_bn.txt", "w", encoding="utf-8") as train_f, \
     open("test_bn.txt", "w", encoding="utf-8") as test_f:

    for line in inp:

        line = line.strip()

        if not line:
            continue

        if random.random() < 0.1:
            test_f.write(line + "\n")
            test_count += 1
        else:
            train_f.write(line + "\n")
            train_count += 1

print("Train:", train_count)
print("Test:", test_count)

Train: 224957
Test: 25043


In [ ]:
import os

print("Train MB:", round(os.path.getsize("train_bn.txt")/1024/1024, 2))
print("Test MB :", round(os.path.getsize("test_bn.txt")/1024/1024, 2))

Train MB: 170.31
Test MB : 18.84


Transliteration

In [ ]:
from aksharamukha import transliterate
import time

start = time.time()

count = 0

with open("train_bn.txt", "r", encoding="utf-8") as fin:
    for line in fin:
        line = line.strip()

        if not line:
            continue

        _ = transliterate.process(
            "Bengali",
            "ISO",
            line
        )

        count += 1

        if count == 1000:
            break

elapsed = time.time() - start

print("1000 docs:", elapsed, "seconds")
print("Docs/sec:", count / elapsed)

1000 docs: 1.8464059829711914 seconds
Docs/sec: 541.5926991261285


In [ ]:
from aksharamukha import transliterate
import time

start = time.time()
count = 0

with open("train_bn.txt", "r", encoding="utf-8") as fin, \
     open("train_iso.txt", "w", encoding="utf-8") as fout:

    for line in fin:

        line = line.strip()

        if not line:
            continue

        iso = transliterate.process(
            "Bengali",
            "HK",
            line
        )

        fout.write(iso + "\n")

        count += 1

        if count % 10000 == 0:
            elapsed = time.time() - start
            print(
                f"{count:,} docs | "
                f"{count/elapsed:.2f} docs/sec"
            )

print("Completed:", count)

10,000 docs | 453.16 docs/sec
20,000 docs | 560.73 docs/sec
30,000 docs | 607.44 docs/sec
40,000 docs | 635.21 docs/sec
50,000 docs | 652.74 docs/sec
60,000 docs | 665.88 docs/sec
70,000 docs | 675.00 docs/sec
80,000 docs | 680.31 docs/sec
90,000 docs | 682.05 docs/sec
100,000 docs | 687.34 docs/sec
110,000 docs | 691.67 docs/sec
120,000 docs | 690.81 docs/sec
130,000 docs | 694.47 docs/sec
140,000 docs | 696.93 docs/sec
150,000 docs | 699.82 docs/sec
160,000 docs | 702.05 docs/sec
170,000 docs | 703.77 docs/sec
180,000 docs | 705.29 docs/sec
190,000 docs | 706.59 docs/sec
200,000 docs | 707.92 docs/sec
210,000 docs | 706.47 docs/sec
220,000 docs | 707.60 docs/sec
Completed: 224957


In [ ]:
from aksharamukha import transliterate
import time

start = time.time()
count = 0

with open("test_bn.txt", "r", encoding="utf-8") as fin, \
     open("test_iso.txt", "w", encoding="utf-8") as fout:

    for line in fin:

        line = line.strip()

        if not line:
            continue

        iso = transliterate.process(
            "Bengali",
            "HK",
            line
        )

        fout.write(iso + "\n")

        count += 1

        if count % 10000 == 0:
            elapsed = time.time() - start
            print(
                f"{count:,} docs | "
                f"{count/elapsed:.2f} docs/sec"
            )

print("Completed:", count)

10,000 docs | 734.27 docs/sec
20,000 docs | 739.73 docs/sec
Completed: 25043


In [ ]:
for fname in [
    "train_bn.txt",
    "train_iso.txt"
]:
    print("\n---", fname, "---")

    with open(fname, "r", encoding="utf-8") as f:
        for _ in range(3):
            print(next(f).strip())


--- train_bn.txt ---
নয়াদিল্লি: আজ ভ্যালেন্টাইনস ডে। এমন একটি দিনে প্রিয় মানুষকে হৃদয় উজাড় করা ভালোবাসার বার্তা দিতে পৌঁছে দিতে উন্মুখ অনেকেই। এমন একটি দিনে ভারতীয় দলের প্রাক্তন ক্রিকেটার সচিন তেন্ডুলকর জানালেন তাঁর প্রথম ভালোবাসার কথা। একটি ভিডিও পোস্ট করেছেন ভারতীয় দলের প্রাক্তন ব্যাটিং স্তম্ভ। ভিডিওতে আন্তর্জাতিক ক্রিকেট ১০০ শতরানের মালিককে নেটে অনুশীলন করতে দেখা যাচ্ছে। ওই ভিডিও শেয়ার করে সচিন লিখেছেন, আমার প্রথম ভালোবাসা।
২য়টা মজার কিন্তু শেষটা ভয়ংকর
বাংলাদেশে এক শ্রেণির লোক আছে যারা প্রসঙ্গ উঠলেই বর্তমান সরকার কে স্বৈরাচার উল্লেখ করে সরকারের আমলে দেশে বাক স্বাধীনতা নাই, কথা বলার স্বাধীনতা নাই বলে সমানতালে জিকির তোলে। অথচ বাক স্বাধীনতার সম্পূর্ণ স্পেস নৈতিক বা অনৈতিক সেটা তারাই ভোগ করে। ব্যাখ্যা চান? বিভিন্ন সভা সেমিনারের বক্তব্য কিংবা টক শো গুলোতে নজর দিন। এহেন এমন কোনো অশোভন বক্তব্য বা মত প্রকাশ বাকি রাখে না তারা, যা উপস্থাপন করে এক হাত নিচ্ছে না। স্বাভাবিক কারণেই জানতে চাইতে পারি, তারপরেও কেন তারা বাক স্বাধীনতা নাই বলে রেগুলার জিকির করে? কেন? জানি এই প্রশ্নের উত্তর না

In [ ]:
import os

print("train_bn.txt :", round(os.path.getsize("train_bn.txt")/1024/1024,2), "MB")
print("train_iso.txt:", round(os.path.getsize("train_iso.txt")/1024/1024,2), "MB")

print("test_bn.txt :", round(os.path.getsize("test_bn.txt")/1024/1024,2), "MB")
print("test_iso.txt:", round(os.path.getsize("test_iso.txt")/1024/1024,2), "MB")

train_bn.txt : 170.31 MB
train_iso.txt: 73.61 MB
test_bn.txt : 18.84 MB
test_iso.txt: 8.14 MB


In [ ]:
from collections import Counter

def corpus_stats(filepath):
    total_chars = 0
    total_words = 0

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()

            if not line:
                continue

            words = line.split()

            total_chars += len(line)
            total_words += len(words)

    avg_word_len = total_chars / total_words

    return {
        "chars": total_chars,
        "words": total_words,
        "avg_word_len": avg_word_len
    }

bn_stats = corpus_stats("train_bn.txt")
ascii_stats = corpus_stats("train_iso.txt")

print("Bengali:", bn_stats)
print("ASCII:", ascii_stats)

Bengali: {'chars': 66913885, 'words': 10178396, 'avg_word_len': 6.5741090246439615}
ASCII: {'chars': 76706745, 'words': 10178231, 'avg_word_len': 7.536353321122305}


In [ ]:
expansion_factor = (
    ascii_stats["chars"] /
    bn_stats["chars"]
)

print("Expansion Factor:", expansion_factor)

Expansion Factor: 1.1463501932371734


In [ ]:
def unique_vocab(filepath):

    vocab = set()

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:

            words = line.split()

            for w in words:
                vocab.add(w)

    return len(vocab)

bn_vocab = unique_vocab("train_bn.txt")
iso_vocab = unique_vocab("train_iso.txt")

print("BN Vocabulary:", bn_vocab)
print("ASCII Vocabulary:", iso_vocab)

BN Vocabulary: 632321
ASCII Vocabulary: 623557


In [ ]:
vocab_ratio = iso_vocab / bn_vocab

print("Vocabulary Ratio:", vocab_ratio)

Vocabulary Ratio: 0.9861399510691563


In [ ]:
import re

bengali_pattern = re.compile(r'[\u0980-\u09FF]')

total_lines = 0
converted_lines = 0

with open("train_iso.txt", "r", encoding="utf-8") as f:

    for line in f:

        total_lines += 1

        if not bengali_pattern.search(line):
            converted_lines += 1

coverage = converted_lines / total_lines

print("Coverage:", coverage)

Coverage: 0.9904292820405678


In [ ]:
char_ratio = (
    ascii_stats["chars"] /
    bn_stats["chars"]
)

print("Character Ratio:", char_ratio)

Character Ratio: 1.1463501932371734


In [ ]:
print("\n=== TRANSLITERATION ANALYSIS ===\n")

print("BN Characters:", bn_stats["chars"])
print("ISO Characters:", ascii_stats["chars"])

print()

print("BN Words:", bn_stats["words"])
print("ISO Words:", ascii_stats["words"])

print()

print("BN Avg Word Length:",
      round(bn_stats["avg_word_len"], 3))

print("ISO Avg Word Length:",
      round(ascii_stats["avg_word_len"], 3))

print()

print("BN Vocabulary:",
      bn_vocab)

print("ISO Vocabulary:",
      iso_vocab)

print()

print("Vocabulary Ratio:",
      round(vocab_ratio, 4))

print("Expansion Factor:",
      round(expansion_factor, 4))

print("Coverage:",
      round(coverage, 4))


=== TRANSLITERATION ANALYSIS ===

BN Characters: 66913885
ISO Characters: 76706745

BN Words: 10178396
ISO Words: 10178231

BN Avg Word Length: 6.574
ISO Avg Word Length: 7.536

BN Vocabulary: 632321
ISO Vocabulary: 623557

Vocabulary Ratio: 0.9861
Expansion Factor: 1.1464
Coverage: 0.9904


In [ ]:
import re

bengali_pattern = re.compile(r'[\u0980-\u09FF]')

count = 0

with open("train_iso.txt", encoding="utf-8") as f:

    for line in f:

        if bengali_pattern.search(line):

            print(line[:300])

            count += 1

            if count == 20:
                break

zanibAra dupure padmA bidhaiৗta jelA rAjabAr3ira goYAlanda bAsasTyAnDe DhAkA-khulanA mahAsar3ake dA~r3iYe kathAgulo balachilena elinA Ahameda (26). tini DhAkAra ekaTi besarakAri bizbabidyAlaYera par3uYA. Aja dupura 12TAra dike goYAlanda bAsasTyAnDe bizrAma neoYA samaYa tA~ra lakSyera biSaYe jAnAna t
AoYAmI lIga prArthI o bartamAna meYara golAma kabirera pakSe mAThe nemechena pirojapura jelA AoYAmI lIgera sabhApati sAbeka saMsada sadasya e ke ema e AuYAla, sAdhAraNa sampAdaka ayADabhokeTa Abdula hAkima hAolAdAra, sAMgaThanika sampAdaka jelA pariSada ceYAramyAna mahiuddina mahArAjasaha zIrSa netA
motAhAra hosena(1897-. )[] paNDita, bAMlAdezera jAtIYa adhyApaka৤ baGkima mukhopAdhyAYa(1897-1961)[65] kamiunisTa netA৤

jale neme kumirera saGge yuddhera prastuti sienaena-era৷

03. Disepazana paYenTa: DyAna brAuna, anu:: mohAmmada nAjima uddina, pR: 352, mUlya: 240৳

bAMlAdeza byAMkera tathyAnuyAYI, calati bacharera sepTembara paryanta moTa bitaraNa karA RNera parimANa 12 lAkha 45 hAjAra 391 k

In [ ]:
import re

# Bengali letters only
letter_pattern = re.compile(r'[অ-হািীুূৃেৈোৌংঃঁ]')

total_lines = 0
clean_lines = 0

with open("train_iso.txt", encoding="utf-8") as f:

    for line in f:

        total_lines += 1

        if not letter_pattern.search(line):
            clean_lines += 1

coverage = clean_lines / total_lines

print("Letter Coverage:", coverage)

Letter Coverage: 1.0


In [ ]:
import re
from collections import Counter

pattern = re.compile(r'[\u0980-\u09FF]')

counter = Counter()

with open("train_iso.txt", encoding="utf-8") as f:

    for line in f:
        counter.update(pattern.findall(line))

print(counter.most_common(20))

[('৷', 8837), ('ৰ', 402), ('\u09e4', 33), ('৳', 31), ('ৱ', 18), ('ৗ', 13), ('৻', 3), ('৺', 3), ('\u09e5', 2), ('৹', 2)]


In [ ]:
bn_ttr = bn_vocab / bn_stats["words"]
iso_ttr = iso_vocab / ascii_stats["words"]

print("BN TTR :", round(bn_ttr, 6))
print("ISO TTR:", round(iso_ttr, 6))

BN TTR : 0.062124
ISO TTR: 0.061264


Tokenization

In [ ]:
!pip -q install tokenizers

In [ ]:
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

tokenizer = Tokenizer(
    BPE(
        unk_token="[UNK]"
    )
)

tokenizer.pre_tokenizer = Whitespace()

trainer = BpeTrainer(
    vocab_size=5000,
    min_frequency=1,
    special_tokens=["[UNK]"]
)

tokenizer.train(
    ["train_bn.txt"],
    trainer
)

tokenizer.save("bpe_bn_5000.json")

print("Vocabulary Size:",
      tokenizer.get_vocab_size())

Vocabulary Size: 5000


In [ ]:
sample = "নয়াদিল্লি: আজ ভ্যালেন্টাইনস ডে।"

enc = tokenizer.encode(sample)

print(enc.tokens)

['ন', '[UNK]', 'া', 'দিল্লি', ':', 'আজ', 'ভ', '্যাল', 'েন্', 'টা', 'ইন', 'স', 'ডে', '।']


In [ ]:
import numpy as np

def evaluate_tokenizer(tokenizer, filepath):

    total_words = 0
    total_subwords = 0
    total_chars = 0

    fragmented_words = 0
    unk_tokens = 0

    doc_token_counts = []

    with open(filepath, "r", encoding="utf-8") as f:

        for line in f:

            line = line.strip()

            if not line:
                continue

            words = line.split()

            doc_tokens = 0

            for word in words:

                total_words += 1
                total_chars += len(word)

                enc = tokenizer.encode(word)

                n_subwords = len(enc.tokens)

                total_subwords += n_subwords
                doc_tokens += n_subwords

                if n_subwords > 1:
                    fragmented_words += 1

                unk_tokens += enc.tokens.count("[UNK]")

            doc_token_counts.append(doc_tokens)

    fertility = total_subwords / total_words

    cpt = total_chars / total_subwords

    compression = total_chars / total_subwords

    oov = unk_tokens / total_subwords

    avg_tokens = np.mean(doc_token_counts)

    wfr = fragmented_words / total_words

    variance = np.var(doc_token_counts)

    return {
        "vocab": tokenizer.get_vocab_size(),
        "oov": oov,
        "fertility": fertility,
        "cpt": cpt,
        "compression": compression,
        "avg_tokens": avg_tokens,
        "wfr": wfr,
        "variance": variance
    }

In [ ]:
print(tokenizer.token_to_id("[UNK]"))

0


In [ ]:
results = evaluate_tokenizer(
    tokenizer,
    "test_bn.txt"
)

print(results)

{'vocab': 5000, 'oov': 6.705718568766416e-05, 'fertility': 1.8265746195200188, 'cpt': 3.061587165837765, 'compression': 3.061587165837765, 'avg_tokens': np.float64(82.1764564948289), 'wfr': 0.48672236477616765, 'variance': np.float64(30236.157219133067)}


In [ ]:
from tokenizers.pre_tokenizers import ByteLevel

In [ ]:
samples = [
    "নয়াদিল্লি",
    "বাংলাদেশ",
    "কলকাতা",
    "ভ্যালেন্টাইনস",
    "ক্রিকেটার"
]

for s in samples:
    print(s)
    print(tokenizer.encode(s).tokens)
    print()

নয়াদিল্লি
['ন', '[UNK]', 'া', 'দিল্লি']

বাংলাদেশ
['বাংলাদেশ']

কলকাতা
['কলকাতা']

ভ্যালেন্টাইনস
['ভ', '্যাল', 'েন্', 'টা', 'ইন', 'স']

ক্রিকেটার
['ক্রিকেটার']



In [ ]:
word = "য়"

print(tokenizer.encode(word).tokens)

['[UNK]']


In [ ]:
print(tokenizer.token_to_id("য়"))

None


In [ ]:
from tokenizers.trainers import BpeTrainer

bengali_chars = list(
    "অআইঈউঊঋএঐওঔ"
    "কখগঘঙচছজঝঞ"
    "টঠডঢণতথদধন"
    "পফবভমযরলশষসহ"
    "ড়ঢ়য়ৎ"
    "ািীুূৃেৈোৌ"
    "ংঃঁ"
    "০১২৩৪৫৬৭৮৯"
    "।"
)

In [ ]:
tokenizer = Tokenizer(
    BPE(unk_token="[UNK]")
)

tokenizer.pre_tokenizer = Whitespace()

tokenizer.train(
    ["train_bn.txt"],
    trainer
)

In [ ]:
trainer = BpeTrainer(
    vocab_size=5000,
    min_frequency=1,
    special_tokens=["[UNK]"],
    initial_alphabet=bengali_chars
)

In [ ]:
for ch in ["য়", "ড়", "ঢ়", "ৎ"]:
    print(ch, tokenizer.encode(ch).tokens)

য় ['[UNK]']
ড় ['[UNK]']
ঢ় ['[UNK]']
ৎ ['ৎ']


In [ ]:
print(tokenizer.get_vocab_size())

5000


In [ ]:
print(tokenizer.token_to_id("য়"))

None


In [ ]:
print(tokenizer.encode("নয়াদিল্লি").tokens)

['ন', '[UNK]', 'া', 'দিল্লি']


In [ ]:
results = evaluate_tokenizer(
    tokenizer,
    "test_bn.txt"
)

print(results)

{'vocab': 5000, 'oov': 6.705718568766416e-05, 'fertility': 1.8265746195200188, 'cpt': 3.061587165837765, 'compression': 3.061587165837765, 'avg_tokens': np.float64(82.1764564948289), 'wfr': 0.48672236477616765, 'variance': np.float64(30236.157219133067)}


In [ ]:
import gc
import csv
import numpy as np

from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

In [ ]:
chars = set()

with open("train_bn.txt", encoding="utf-8") as f:
    for line in f:
        chars.update(line)

chars.discard(" ")
chars.discard("\n")

bengali_chars = sorted(list(chars))

print("Character inventory:", len(bengali_chars))

Character inventory: 1435


In [ ]:
def evaluate_tokenizer(tokenizer, filepath):

    total_words = 0
    total_subwords = 0
    total_chars = 0

    fragmented_words = 0
    unk_tokens = 0

    doc_token_counts = []

    with open(filepath, "r", encoding="utf-8") as f:

        for line in f:

            line = line.strip()

            if not line:
                continue

            words = line.split()

            doc_tokens = 0

            for word in words:

                total_words += 1
                total_chars += len(word)

                enc = tokenizer.encode(word)

                n_subwords = len(enc.tokens)

                total_subwords += n_subwords
                doc_tokens += n_subwords

                if n_subwords > 1:
                    fragmented_words += 1

                unk_tokens += enc.tokens.count("[UNK]")

            doc_token_counts.append(doc_tokens)

    fertility = total_subwords / total_words

    cpt = total_chars / total_subwords

    compression = total_chars / total_subwords

    oov = unk_tokens / total_subwords

    avg_tokens = np.mean(doc_token_counts)

    wfr = fragmented_words / total_words

    variance = np.var(doc_token_counts)

    return {
        "vocab": tokenizer.get_vocab_size(),
        "oov": oov,
        "fertility": fertility,
        "cpt": cpt,
        "compression": compression,
        "avg_tokens": avg_tokens,
        "wfr": wfr,
        "variance": variance
    }

In [ ]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

results_file = "bpe_bn_results.csv"

with open(results_file, "w", newline="") as f:

    writer = csv.writer(f)

    writer.writerow([
        "script",
        "tokenizer",
        "vocab",
        "oov",
        "fertility",
        "cpt",
        "compression",
        "avg_tokens",
        "wfr",
        "variance"
    ])

for vocab_size in VOCABS:

    print(f"\nTraining BPE-{vocab_size}")

    tokenizer = Tokenizer(
        BPE(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = Whitespace()

    trainer = BpeTrainer(
        vocab_size=vocab_size,
        min_frequency=1,
        special_tokens=["[UNK]"],
        initial_alphabet=bengali_chars
    )

    tokenizer.train(
        ["train_bn.txt"],
        trainer
    )

    tokenizer.save(
        f"bpe_bn_{vocab_size}.json"
    )

    metrics = evaluate_tokenizer(
        tokenizer,
        "test_bn.txt"
    )

    print(metrics)

    with open(results_file, "a", newline="") as f:

        writer = csv.writer(f)

        writer.writerow([
            "Bengali",
            "BPE",
            metrics["vocab"],
            metrics["oov"],
            metrics["fertility"],
            metrics["cpt"],
            metrics["compression"],
            metrics["avg_tokens"],
            metrics["wfr"],
            metrics["variance"]
        ])

    del tokenizer
    gc.collect()

print("\nFinished.")


Training BPE-5000
{'vocab': 5000, 'oov': 6.705718568766416e-05, 'fertility': 1.8265746195200188, 'cpt': 3.061587165837765, 'compression': 3.061587165837765, 'avg_tokens': np.float64(82.1764564948289), 'wfr': 0.48672236477616765, 'variance': np.float64(30236.157219133067)}

Training BPE-10000
{'vocab': 10000, 'oov': 7.916098533628791e-05, 'fertility': 1.5472893991047947, 'cpt': 3.6142026280299873, 'compression': 3.6142026280299873, 'avg_tokens': np.float64(69.61158806852214), 'wfr': 0.359959313693729, 'variance': np.float64(21921.07458839366)}

Training BPE-15000
{'vocab': 15000, 'oov': 8.492715342644139e-05, 'fertility': 1.4422354746602595, 'cpt': 3.877464887545371, 'compression': 3.877464887545371, 'avg_tokens': np.float64(64.88527732300443), 'wfr': 0.3050434510934445, 'variance': np.float64(19086.04509850054)}

Training BPE-20000
{'vocab': 20000, 'oov': 8.834849769141534e-05, 'fertility': 1.3863841110388233, 'cpt': 4.033671020924509, 'compression': 4.033671020924509, 'avg_tokens': n

In [ ]:
with open("bpe_bn_results.csv", encoding="utf-8") as f:
    print(f.read())

script,tokenizer,vocab,oov,fertility,cpt,compression,avg_tokens,wfr,variance
Bengali,BPE,5000,6.705718568766416e-05,1.8265746195200188,3.061587165837765,3.061587165837765,82.1764564948289,0.48672236477616765,30236.157219133067
Bengali,BPE,10000,7.916098533628791e-05,1.5472893991047947,3.6142026280299873,3.6142026280299873,69.61158806852214,0.359959313693729,21921.07458839366
Bengali,BPE,15000,8.492715342644139e-05,1.4422354746602595,3.877464887545371,3.877464887545371,64.88527732300443,0.3050434510934445,19086.04509850054
Bengali,BPE,20000,8.834849769141534e-05,1.3863841110388233,4.033671020924509,4.033671020924509,62.37255919817913,0.27487221180311167,17591.76288873861
Bengali,BPE,25000,9.06612953134023e-05,1.3510170245209552,4.139264947124099,4.139264947124099,60.78141596454099,0.25485746035437207,16656.31224178373
Bengali,BPE,30000,9.236341926700926e-05,1.3261197388052746,4.216977735061556,4.216977735061556,59.661302559597495,0.24066606962648301,16000.675125516533
Bengali,BPE,35000,

In [ ]:
chars = set()

with open("train_iso.txt", encoding="utf-8") as f:
    for line in f:
        chars.update(line)

chars.discard(" ")
chars.discard("\n")

iso_chars = sorted(list(chars))

print("Character inventory:", len(iso_chars))
print(iso_chars[:50])

Character inventory: 1351
['!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R']


In [ ]:
import gc
import csv

from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

results_file = "bpe_iso_results.csv"

with open(results_file, "w", newline="") as f:

    writer = csv.writer(f)

    writer.writerow([
        "script",
        "tokenizer",
        "vocab",
        "oov",
        "fertility",
        "cpt",
        "compression",
        "avg_tokens",
        "wfr",
        "variance"
    ])

for vocab_size in VOCABS:

    print(f"\nTraining ISO BPE-{vocab_size}")

    tokenizer = Tokenizer(
        BPE(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = Whitespace()

    trainer = BpeTrainer(
        vocab_size=vocab_size,
        min_frequency=1,
        special_tokens=["[UNK]"],
        initial_alphabet=iso_chars
    )

    tokenizer.train(
        ["train_iso.txt"],
        trainer
    )

    tokenizer.save(
        f"bpe_iso_{vocab_size}.json"
    )

    metrics = evaluate_tokenizer(
        tokenizer,
        "test_iso.txt"
    )

    print(metrics)

    with open(results_file, "a", newline="") as f:

        writer = csv.writer(f)

        writer.writerow([
            "ISO",
            "BPE",
            metrics["vocab"],
            metrics["oov"],
            metrics["fertility"],
            metrics["cpt"],
            metrics["compression"],
            metrics["avg_tokens"],
            metrics["wfr"],
            metrics["variance"]
        ])

    del tokenizer
    gc.collect()

print("\nFinished.")


Training ISO BPE-5000
{'vocab': 5000, 'oov': 6.755189013125626e-05, 'fertility': 1.8132173323380585, 'cpt': 3.6132213734180376, 'compression': 3.6132213734180376, 'avg_tokens': np.float64(81.5746515992493), 'wfr': 0.48377900283759834, 'variance': np.float64(29682.347849252696)}

Training ISO BPE-10000
{'vocab': 10000, 'oov': 7.920201104236734e-05, 'fertility': 1.5465043930850295, 'cpt': 4.236364053765539, 'compression': 4.236364053765539, 'avg_tokens': np.float64(69.57553008824821), 'wfr': 0.360378535792171, 'variance': np.float64(21873.623922247254)}

Training ISO BPE-15000
{'vocab': 15000, 'oov': 8.472026963146683e-05, 'fertility': 1.4457727595887657, 'cpt': 4.531525149027866, 'compression': 4.531525149027866, 'avg_tokens': np.float64(65.04372479335542), 'wfr': 0.30722305013859585, 'variance': np.float64(19136.015458265832)}

Training ISO BPE-20000
{'vocab': 20000, 'oov': 8.796689385073165e-05, 'fertility': 1.392413130171827, 'cpt': 4.705180867583052, 'compression': 4.70518086758305

In [ ]:
with open("bpe_iso_results.csv", encoding="utf-8") as f:
    print(f.read())

script,tokenizer,vocab,oov,fertility,cpt,compression,avg_tokens,wfr,variance
ISO,BPE,5000,6.755189013125626e-05,1.8132173323380585,3.6132213734180376,3.6132213734180376,81.5746515992493,0.48377900283759834,29682.347849252696
ISO,BPE,10000,7.920201104236734e-05,1.5465043930850295,4.236364053765539,4.236364053765539,69.57553008824821,0.360378535792171,21873.623922247254
ISO,BPE,15000,8.472026963146683e-05,1.4457727595887657,4.531525149027866,4.531525149027866,65.04372479335542,0.30722305013859585,19136.015458265832
ISO,BPE,20000,8.796689385073165e-05,1.392413130171827,4.705180867583052,4.705180867583052,62.643133809847065,0.2777349273115065,17705.694353186864
ISO,BPE,25000,9.018022321565685e-05,1.358238576603172,4.823567621117594,4.823567621117594,61.10565826777942,0.2579658227836866,16812.31539465014
ISO,BPE,30000,9.17764936458079e-05,1.3346147052740984,4.908949072691638,4.908949072691638,60.04284630435651,0.24432191873835604,16185.69009407481
ISO,BPE,35000,9.295345456798872e-05,1.31771

In [ ]:
import csv

bn = []
iso = []

with open("bpe_bn_results.csv") as f:
    bn = list(csv.DictReader(f))

with open("bpe_iso_results.csv") as f:
    iso = list(csv.DictReader(f))

print(
    f"{'Vocab':<8}"
    f"{'BN_Fert':<12}"
    f"{'ISO_Fert':<12}"
    f"{'BN_CPT':<12}"
    f"{'ISO_CPT':<12}"
)

for b, i in zip(bn, iso):

    print(
        f"{b['vocab']:<8}"
        f"{float(b['fertility']):<12.4f}"
        f"{float(i['fertility']):<12.4f}"
        f"{float(b['cpt']):<12.4f}"
        f"{float(i['cpt']):<12.4f}"
    )

Vocab   BN_Fert     ISO_Fert    BN_CPT      ISO_CPT     
5000    1.8266      1.8132      3.0616      3.6132      
10000   1.5473      1.5465      3.6142      4.2364      
15000   1.4422      1.4458      3.8775      4.5315      
20000   1.3864      1.3924      4.0337      4.7052      
25000   1.3510      1.3582      4.1393      4.8236      
30000   1.3261      1.3346      4.2170      4.9089      
35000   1.3080      1.3177      4.2754      4.9719      
40000   1.2943      1.3045      4.3207      5.0224      
45000   1.2833      1.2942      4.3577      5.0623      
50000   1.2745      1.2858      4.3877      5.0952      


In [ ]:
import gc
import csv

from tokenizers import Tokenizer
from tokenizers.models import WordPiece
from tokenizers.trainers import WordPieceTrainer
from tokenizers.pre_tokenizers import Whitespace

In [ ]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

results_file = "wp_bn_results.csv"

with open(results_file, "w", newline="") as f:

    writer = csv.writer(f)

    writer.writerow([
        "script",
        "tokenizer",
        "vocab",
        "oov",
        "fertility",
        "cpt",
        "compression",
        "avg_tokens",
        "wfr",
        "variance"
    ])

for vocab_size in VOCABS:

    print(f"\nTraining Bengali WordPiece-{vocab_size}")

    tokenizer = Tokenizer(
        WordPiece(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = Whitespace()

    trainer = WordPieceTrainer(
        vocab_size=vocab_size,
        min_frequency=1,
        special_tokens=["[UNK]"],
        initial_alphabet=bengali_chars
    )

    tokenizer.train(
        ["train_bn.txt"],
        trainer
    )

    tokenizer.save(
        f"wp_bn_{vocab_size}.json"
    )

    metrics = evaluate_tokenizer(
        tokenizer,
        "test_bn.txt"
    )

    print(metrics)

    with open(results_file, "a", newline="") as f:

        writer = csv.writer(f)

        writer.writerow([
            "Bengali",
            "WordPiece",
            metrics["vocab"],
            metrics["oov"],
            metrics["fertility"],
            metrics["cpt"],
            metrics["compression"],
            metrics["avg_tokens"],
            metrics["wfr"],
            metrics["variance"]
        ])

    del tokenizer
    gc.collect()

print("\nFinished.")


Training Bengali WordPiece-5000
{'vocab': 5000, 'oov': 3.5178940684873916e-05, 'fertility': 2.0688773721474543, 'cpt': 2.703020240761238, 'compression': 2.703020240761238, 'avg_tokens': np.float64(93.07750668849579), 'wfr': 0.5675757476241913, 'variance': np.float64(38447.318314879114)}

Training Bengali WordPiece-10000
{'vocab': 10000, 'oov': 4.4429730858777934e-05, 'fertility': 1.6381128796478823, 'cpt': 3.4138168877406994, 'compression': 3.4138168877406994, 'avg_tokens': np.float64(73.69767999041649), 'wfr': 0.39637551046491915, 'variance': np.float64(24549.052914395543)}

Training Bengali WordPiece-15000
{'vocab': 15000, 'oov': 4.854926507069839e-05, 'fertility': 1.499114646803986, 'cpt': 3.730346724641594, 'compression': 3.730346724641594, 'avg_tokens': np.float64(67.44423591422753), 'wfr': 0.3295493175014135, 'variance': np.float64(20647.21115183701)}

Training Bengali WordPiece-20000
{'vocab': 20000, 'oov': 5.098113822230015e-05, 'fertility': 1.4276047357298371, 'cpt': 3.917202

In [ ]:
with open("wp_bn_results.csv", encoding="utf-8") as f:
    print(f.read())

script,tokenizer,vocab,oov,fertility,cpt,compression,avg_tokens,wfr,variance
Bengali,WordPiece,5000,3.5178940684873916e-05,2.0688773721474543,2.703020240761238,2.703020240761238,93.07750668849579,0.5675757476241913,38447.318314879114
Bengali,WordPiece,10000,4.4429730858777934e-05,1.6381128796478823,3.4138168877406994,3.4138168877406994,73.69767999041649,0.39637551046491915,24549.052914395543
Bengali,WordPiece,15000,4.854926507069839e-05,1.499114646803986,3.730346724641594,3.730346724641594,67.44423591422753,0.3295493175014135,20647.21115183701
Bengali,WordPiece,20000,5.098113822230015e-05,1.4276047357298371,3.9172029011997975,3.9172029011997975,64.22704947490317,0.29316152303826587,18697.770035406527
Bengali,WordPiece,25000,5.261270186114225e-05,1.3833335256406274,4.042566242278926,4.042566242278926,62.23531525775666,0.2697739975094726,17527.696213999363
Bengali,WordPiece,30000,5.38079280076171e-05,1.3526057786270858,4.134403017443612,4.134403017443612,60.852893023998725,0.253023736341

In [ ]:
import gc
import csv

from tokenizers import Tokenizer
from tokenizers.models import WordPiece
from tokenizers.trainers import WordPieceTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

results_file = "wp_iso_results.csv"

with open(results_file, "w", newline="") as f:

    writer = csv.writer(f)

    writer.writerow([
        "script",
        "tokenizer",
        "vocab",
        "oov",
        "fertility",
        "cpt",
        "compression",
        "avg_tokens",
        "wfr",
        "variance"
    ])

for vocab_size in VOCABS:

    print(f"\nTraining ISO WordPiece-{vocab_size}")

    tokenizer = Tokenizer(
        WordPiece(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = Whitespace()

    trainer = WordPieceTrainer(
        vocab_size=vocab_size,
        min_frequency=1,
        special_tokens=["[UNK]"],
        initial_alphabet=iso_chars
    )

    tokenizer.train(
        ["train_iso.txt"],
        trainer
    )

    tokenizer.save(
        f"wp_iso_{vocab_size}.json"
    )

    metrics = evaluate_tokenizer(
        tokenizer,
        "test_iso.txt"
    )

    print(metrics)

    with open(results_file, "a", newline="") as f:

        writer = csv.writer(f)

        writer.writerow([
            "ISO",
            "WordPiece",
            metrics["vocab"],
            metrics["oov"],
            metrics["fertility"],
            metrics["cpt"],
            metrics["compression"],
            metrics["avg_tokens"],
            metrics["wfr"],
            metrics["variance"]
        ])

    del tokenizer
    gc.collect()

print("\nFinished.")


Training ISO WordPiece-5000
{'vocab': 5000, 'oov': 3.56604574506296e-05, 'fertility': 2.0658532277347943, 'cpt': 3.1713558019993915, 'compression': 3.1713558019993915, 'avg_tokens': np.float64(92.94046240466398), 'wfr': 0.5640563188264042, 'variance': np.float64(38218.98335780239)}

Training ISO WordPiece-10000
{'vocab': 10000, 'oov': 4.464280911918124e-05, 'fertility': 1.6501934484053264, 'cpt': 3.970174300587026, 'compression': 3.970174300587026, 'avg_tokens': np.float64(74.24038653515953), 'wfr': 0.39915608743388625, 'variance': np.float64(24832.7272640282)}

Training ISO WordPiece-15000
{'vocab': 15000, 'oov': 4.871798333375399e-05, 'fertility': 1.5121576486898851, 'cpt': 4.33258769383741, 'compression': 4.33258769383741, 'avg_tokens': np.float64(68.0303078704628), 'wfr': 0.3322182350085252, 'variance': np.float64(20938.979275499194)}

Training ISO WordPiece-20000
{'vocab': 20000, 'oov': 5.113898219870514e-05, 'fertility': 1.4405697563677322, 'cpt': 4.547891964895245, 'compression

In [ ]:
!pip -q install sentencepiece

In [ ]:
import sentencepiece as spm

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

for vocab_size in VOCABS:

    print(f"\nTraining Unigram-{vocab_size}")

    spm.SentencePieceTrainer.train(
        input="train_bn.txt",
        model_prefix=f"uni_bn_{vocab_size}",
        vocab_size=vocab_size,
        model_type="unigram",
        character_coverage=1.0,
        unk_id=0,
        bos_id=-1,
        eos_id=-1,
        pad_id=-1,
        minloglevel=1
    )

print("Finished")


Training Unigram-5000

Training Unigram-10000

Training Unigram-15000

Training Unigram-20000

Training Unigram-25000

Training Unigram-30000

Training Unigram-35000

Training Unigram-40000

Training Unigram-45000

Training Unigram-50000
Finished


In [ ]:
import sentencepiece as spm
import numpy as np

def evaluate_unigram(model_path, filepath):

    sp = spm.SentencePieceProcessor()
    sp.load(model_path)

    total_words = 0
    total_subwords = 0
    total_chars = 0

    fragmented_words = 0
    unk_tokens = 0

    doc_token_counts = []

    with open(filepath, encoding="utf-8") as f:

        for line in f:

            words = line.strip().split()

            doc_tokens = 0

            for word in words:

                total_words += 1
                total_chars += len(word)

                pieces = sp.encode(word, out_type=str)

                n_subwords = len(pieces)

                total_subwords += n_subwords
                doc_tokens += n_subwords

                if n_subwords > 1:
                    fragmented_words += 1

                unk_tokens += pieces.count("<unk>")

            doc_token_counts.append(doc_tokens)

    return {
        "vocab": sp.get_piece_size(),
        "oov": unk_tokens / total_subwords,
        "fertility": total_subwords / total_words,
        "cpt": total_chars / total_subwords,
        "compression": total_chars / total_subwords,
        "avg_tokens": np.mean(doc_token_counts),
        "wfr": fragmented_words / total_words,
        "variance": np.var(doc_token_counts)
    }

In [ ]:
import csv

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

with open("uni_bn_results.csv", "w", newline="") as f:

    writer = csv.writer(f)

    writer.writerow([
        "script",
        "tokenizer",
        "vocab",
        "oov",
        "fertility",
        "cpt",
        "compression",
        "avg_tokens",
        "wfr",
        "variance"
    ])

for vocab_size in VOCABS:

    metrics = evaluate_unigram(
        f"uni_bn_{vocab_size}.model",
        "test_bn.txt"
    )

    print(metrics)

    with open("uni_bn_results.csv", "a", newline="") as f:

        writer = csv.writer(f)

        writer.writerow([
            "Bengali",
            "Unigram",
            metrics["vocab"],
            metrics["oov"],
            metrics["fertility"],
            metrics["cpt"],
            metrics["compression"],
            metrics["avg_tokens"],
            metrics["wfr"],
            metrics["variance"]
        ])

print("Finished")

{'vocab': 5000, 'oov': 0.0, 'fertility': 1.8507467588084876, 'cpt': 3.0216005436463726, 'compression': 3.0216005436463726, 'avg_tokens': np.float64(83.26394601285789), 'wfr': 0.4541529056004914, 'variance': np.float64(31798.44656458312)}
{'vocab': 10000, 'oov': 0.0, 'fertility': 1.5779443652039773, 'cpt': 3.543988961768367, 'compression': 3.543988961768367, 'avg_tokens': np.float64(70.99073593419318), 'wfr': 0.3534161319784249, 'variance': np.float64(23360.101659175692)}
{'vocab': 15000, 'oov': 0.0, 'fertility': 1.4745706147945847, 'cpt': 3.7924378503545, 'compression': 3.7924378503545, 'avg_tokens': np.float64(66.3400151739009), 'wfr': 0.30715764789836236, 'variance': np.float64(20361.707454010873)}
{'vocab': 20000, 'oov': 0.0, 'fertility': 1.4180260573424848, 'cpt': 3.943663364725352, 'compression': 3.943663364725352, 'avg_tokens': np.float64(63.79611068961386), 'wfr': 0.28119705077533863, 'variance': np.float64(18802.604837327046)}
{'vocab': 25000, 'oov': 0.0, 'fertility': 1.3825160

In [ ]:
with open("uni_bn_results.csv", encoding="utf-8") as f:
    print(f.read())

In [ ]:
import sentencepiece as spm

VOCABS = [
    5000,10000,15000,20000,25000,
    30000,35000,40000,45000,50000
]

for vocab_size in VOCABS:

    print(f"\nTraining ISO Unigram-{vocab_size}")

    spm.SentencePieceTrainer.train(
        input="train_iso.txt",
        model_prefix=f"uni_iso_{vocab_size}",
        vocab_size=vocab_size,
        model_type="unigram",
        character_coverage=1.0,
        unk_id=0,
        bos_id=-1,
        eos_id=-1,
        pad_id=-1,
        minloglevel=1
    )

print("Finished")


Training ISO Unigram-5000

Training ISO Unigram-10000

Training ISO Unigram-15000

Training ISO Unigram-20000

Training ISO Unigram-25000

Training ISO Unigram-30000

Training ISO Unigram-35000

Training ISO Unigram-40000

Training ISO Unigram-45000

Training ISO Unigram-50000
Finished


In [ ]:
import csv

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

with open("uni_iso_results.csv", "w", newline="") as f:

    writer = csv.writer(f)

    writer.writerow([
        "script",
        "tokenizer",
        "vocab",
        "oov",
        "fertility",
        "cpt",
        "compression",
        "avg_tokens",
        "wfr",
        "variance"
    ])

for vocab_size in VOCABS:

    metrics = evaluate_unigram(
        f"uni_iso_{vocab_size}.model",
        "test_iso.txt"
    )

    print(metrics)

    with open("uni_iso_results.csv", "a", newline="") as f:

        writer = csv.writer(f)

        writer.writerow([
            "Bengali",
            "Unigram",
            metrics["vocab"],
            metrics["oov"],
            metrics["fertility"],
            metrics["cpt"],
            metrics["compression"],
            metrics["avg_tokens"],
            metrics["wfr"],
            metrics["variance"]
        ])

print("Finished")

{'vocab': 5000, 'oov': 0.0, 'fertility': 1.9291452500627964, 'cpt': 3.3960924506035237, 'compression': 3.3960924506035237, 'avg_tokens': np.float64(86.79012099189394), 'wfr': 0.4620785207920423, 'variance': np.float64(34133.06260735269)}
{'vocab': 10000, 'oov': 0.0, 'fertility': 1.6522473121810808, 'cpt': 3.9652390847030055, 'compression': 3.9652390847030055, 'avg_tokens': np.float64(74.33278760531886), 'wfr': 0.3641436568538606, 'variance': np.float64(25424.694028195394)}
{'vocab': 15000, 'oov': 0.0, 'fertility': 1.5441327751036917, 'cpt': 4.242870642659573, 'compression': 4.242870642659573, 'avg_tokens': np.float64(69.46883360619734), 'wfr': 0.32191607561129965, 'variance': np.float64(22206.339592885426)}
{'vocab': 20000, 'oov': 0.0, 'fertility': 1.4855248758051474, 'cpt': 4.410263151133702, 'compression': 4.410263151133702, 'avg_tokens': np.float64(66.83212873856966), 'wfr': 0.29827090232430986, 'variance': np.float64(20491.005481340773)}
{'vocab': 25000, 'oov': 0.0, 'fertility': 1.


# Table 1. BPE Results (Bangla)

|  Vocab |      OOV |  Fertility |        CPT | Avg. Tokens |        WFR |     Variance |
| -----: | -------: | ---------: | ---------: | ----------: | ---------: | -----------: |
|  5,000 | 0.000067 |     1.8266 |     3.0616 |     82.1765 |     0.4867 |     30236.16 |
| 10,000 | 0.000079 |     1.5473 |     3.6142 |     69.6116 |     0.3600 |     21921.07 |
| 15,000 | 0.000085 |     1.4422 |     3.8775 |     64.8853 |     0.3050 |     19086.05 |
| 20,000 | 0.000088 |     1.3864 |     4.0337 |     62.3726 |     0.2749 |     17591.76 |
| 25,000 | 0.000091 |     1.3510 |     4.1393 |     60.7814 |     0.2549 |     16656.31 |
| 30,000 | 0.000092 |     1.3261 |     4.2170 |     59.6613 |     0.2407 |     16000.68 |
| 35,000 | 0.000094 |     1.3080 |     4.2754 |     58.8456 |     0.2301 |     15537.40 |
| 40,000 | 0.000095 |     1.2943 |     4.3207 |     58.2294 |     0.2219 |     15201.74 |
| 45,000 | 0.000095 |     1.2833 |     4.3577 |     57.7348 |     0.2154 |     14914.70 |
| 50,000 | 0.000096 | **1.2745** | **4.3877** | **57.3395** | **0.2102** | **14680.40** |

---

# Table 2. BPE Results (ASCII)

|  Vocab |      OOV |  Fertility |        CPT | Avg. Tokens |        WFR |     Variance |
| -----: | -------: | ---------: | ---------: | ----------: | ---------: | -----------: |
|  5,000 | 0.000068 |     1.8132 |     3.6132 |     81.5747 |     0.4838 |     29682.35 |
| 10,000 | 0.000079 |     1.5465 |     4.2364 |     69.5755 |     0.3604 |     21873.62 |
| 15,000 | 0.000085 |     1.4458 |     4.5315 |     65.0437 |     0.3072 |     19136.02 |
| 20,000 | 0.000088 |     1.3924 |     4.7052 |     62.6431 |     0.2777 |     17705.69 |
| 25,000 | 0.000090 |     1.3582 |     4.8236 |     61.1057 |     0.2580 |     16812.32 |
| 30,000 | 0.000092 |     1.3346 |     4.9089 |     60.0428 |     0.2443 |     16185.69 |
| 35,000 | 0.000093 |     1.3177 |     4.9719 |     59.2826 |     0.2342 |     15753.47 |
| 40,000 | 0.000094 |     1.3045 |     5.0224 |     58.6870 |     0.2262 |     15409.39 |
| 45,000 | 0.000095 |     1.2942 |     5.0623 |     58.2235 |     0.2200 |     15144.32 |
| 50,000 | 0.000095 | **1.2858** | **5.0952** | **57.8477** | **0.2149** | **14922.79** |

---

# Table 3. Best BPE Comparison (50k Vocabulary)

| Metric      |       Bangla |        ASCII | Better             |
| ----------- | -----------: | -----------: | :----------------- |
| OOV         |     0.000096 | **0.000095** | ASCII (negligible) |
| Fertility   |   **1.2745** |       1.2858 | **Bangla**         |
| CPT         |       4.3877 |   **5.0952** | ASCII*             |
| Avg. Tokens |  **57.3395** |      57.8477 | **Bangla**         |
| WFR         |   **0.2102** |       0.2149 | **Bangla**         |
| Variance    | **14680.40** |     14922.79 | **Bangla**         |

Higher CPT is expected for ASCII because transliterated words contain more Latin characters per token and does not necessarily indicate superior segmentation.



## Table 4. WordPiece Results (Bangla)

|  Vocab |      OOV |  Fertility |        CPT | Avg. Tokens |        WFR |     Variance |
| -----: | -------: | ---------: | ---------: | ----------: | ---------: | -----------: |
|  5,000 | 0.000035 |     2.0689 |     2.7030 |     93.0775 |     0.5676 |     38447.32 |
| 10,000 | 0.000044 |     1.6381 |     3.4138 |     73.6977 |     0.3964 |     24549.05 |
| 15,000 | 0.000049 |     1.4991 |     3.7303 |     67.4442 |     0.3295 |     20647.21 |
| 20,000 | 0.000051 |     1.4276 |     3.9172 |     64.2270 |     0.2932 |     18697.77 |
| 25,000 | 0.000053 |     1.3833 |     4.0426 |     62.2353 |     0.2698 |     17527.70 |
| 30,000 | 0.000054 |     1.3526 |     4.1344 |     60.8529 |     0.2530 |     16734.56 |
| 35,000 | 0.000055 |     1.3302 |     4.2040 |     59.8456 |     0.2407 |     16149.79 |
| 40,000 | 0.000055 |     1.3135 |     4.2575 |     59.0930 |     0.2313 |     15696.74 |
| 45,000 | 0.000056 |     1.3004 |     4.3005 |     58.5032 |     0.2238 |     15373.75 |
| 50,000 | 0.000056 | **1.2896** | **4.3363** | **58.0200** | **0.2177** | **15091.11** |

---

## Table 5. WordPiece Results (ASCII)

|  Vocab |      OOV |  Fertility |        CPT | Avg. Tokens |        WFR |     Variance |
| -----: | -------: | ---------: | ---------: | ----------: | ---------: | -----------: |
|  5,000 | 0.000036 |     2.0659 |     3.1714 |     92.9405 |     0.5641 |     38218.98 |
| 10,000 | 0.000045 |     1.6502 |     3.9702 |     74.2404 |     0.3992 |     24832.73 |
| 15,000 | 0.000049 |     1.5122 |     4.3326 |     68.0303 |     0.3322 |     20938.98 |
| 20,000 | 0.000051 |     1.4406 |     4.5479 |     64.8096 |     0.2966 |     18989.98 |
| 25,000 | 0.000053 |     1.3959 |     4.6933 |     62.8013 |     0.2736 |     17813.44 |
| 30,000 | 0.000054 |     1.3656 |     4.7976 |     61.4360 |     0.2571 |     17043.92 |
| 35,000 | 0.000055 |     1.3434 |     4.8769 |     60.4378 |     0.2453 |     16436.90 |
| 40,000 | 0.000056 |     1.3269 |     4.9374 |     59.6968 |     0.2360 |     16000.91 |
| 45,000 | 0.000056 |     1.3136 |     4.9873 |     59.0993 |     0.2286 |     15658.77 |
| 50,000 | 0.000057 | **1.3031** | **5.0277** | **58.6248** | **0.2225** | **15391.12** |

---

## Table 6. Best WordPiece Comparison (50k Vocabulary)

| Metric      |       Bangla |      ASCII | Better                             |
| ----------- | -----------: | ---------: | :--------------------------------- |
| OOV         | **0.000056** |   0.000057 | **Bangla** (difference negligible) |
| Fertility   |   **1.2896** |     1.3031 | **Bangla**                         |
| CPT         |       4.3363 | **5.0277** | ASCII*                             |
| Avg. Tokens |  **58.0200** |    58.6248 | **Bangla**                         |
| WFR         |   **0.2177** |     0.2225 | **Bangla**                         |
| Variance    | **15091.11** |   15391.12 | **Bangla**                         |

Higher CPT for ASCII is expected because transliterated words contain more Latin characters per token; it reflects the script representation rather than better segmentation quality.


## Table 7. SentencePiece Unigram Results (Bangla)

|  Vocab |      OOV |  Fertility |        CPT | Avg. Tokens |        WFR |     Variance |
| -----: | -------: | ---------: | ---------: | ----------: | ---------: | -----------: |
|  5,000 | 0.000000 |     1.8507 |     3.0216 |     83.2639 |     0.4542 |     31798.45 |
| 10,000 | 0.000000 |     1.5779 |     3.5440 |     70.9907 |     0.3534 |     23360.10 |
| 15,000 | 0.000000 |     1.4746 |     3.7924 |     66.3400 |     0.3072 |     20361.71 |
| 20,000 | 0.000000 |     1.4180 |     3.9437 |     63.7961 |     0.2812 |     18802.60 |
| 25,000 | 0.000000 |     1.3825 |     4.0450 |     62.1985 |     0.2648 |     17812.11 |
| 30,000 | 0.000000 |     1.3586 |     4.1163 |     61.1207 |     0.2531 |     17157.81 |
| 35,000 | 0.000000 |     1.3408 |     4.1709 |     60.3201 |     0.2451 |     16647.59 |
| 40,000 | 0.000000 |     1.3281 |     4.2106 |     59.7519 |     0.2394 |     16328.78 |
| 45,000 | 0.000000 |     1.3177 |     4.2440 |     59.2815 |     0.2347 |     16033.63 |
| 50,000 | 0.000000 | **1.3097** | **4.2699** | **58.9222** | **0.2310** | **15773.49** |

---

## Table 8. SentencePiece Unigram Results (ASCII)

|  Vocab |      OOV |  Fertility |        CPT | Avg. Tokens |        WFR |     Variance |
| -----: | -------: | ---------: | ---------: | ----------: | ---------: | -----------: |
|  5,000 | 0.000000 |     1.9291 |     3.3961 |     86.7901 |     0.4621 |     34133.06 |
| 10,000 | 0.000000 |     1.6522 |     3.9652 |     74.3328 |     0.3641 |     25424.69 |
| 15,000 | 0.000000 |     1.5441 |     4.2429 |     69.4688 |     0.3219 |     22206.34 |
| 20,000 | 0.000000 |     1.4855 |     4.4103 |     66.8321 |     0.2983 |     20491.01 |
| 25,000 | 0.000000 |     1.4488 |     4.5220 |     65.1806 |     0.2838 |     19407.52 |
| 30,000 | 0.000000 |     1.4234 |     4.6028 |     64.0366 |     0.2738 |     18718.90 |
| 35,000 | 0.000000 |     1.4055 |     4.6615 |     63.2302 |     0.2666 |     18185.60 |
| 40,000 | 0.000000 |     1.3923 |     4.7057 |     62.6369 |     0.2614 |     17824.16 |
| 45,000 | 0.000000 |     1.3822 |     4.7399 |     62.1846 |     0.2573 |     17541.88 |
| 50,000 | 0.000000 | **1.3736** | **4.7695** | **61.7988** | **0.2540** | **17306.75** |

---

## Table 9. Best SentencePiece Unigram Comparison (50k Vocabulary)

| Metric      |       Bangla |        ASCII | Better     |
| ----------- | -----------: | -----------: | :--------- |
| OOV         | **0.000000** | **0.000000** | Tie        |
| Fertility   |   **1.3097** |       1.3736 | **Bangla** |
| CPT         |       4.2699 |   **4.7695** | ASCII*     |
| Avg. Tokens |  **58.9222** |      61.7988 | **Bangla** |
| WFR         |   **0.2310** |       0.2540 | **Bangla** |
| Variance    | **15773.49** |     17306.75 | **Bangla** |

Higher CPT for ASCII is expected because transliterated words contain more Latin characters per token and does not indicate superior tokenization quality.

---

## Table 10. Overall Best Tokenizer Comparison (50k Vocabulary)

| Tokenizer                 | Script |          OOV |  Fertility |    CPT | Avg. Tokens |        WFR |     Variance |
| ------------------------- | ------ | -----------: | ---------: | -----: | ----------: | ---------: | -----------: |
| **BPE**                   | Bangla |     0.000096 | **1.2745** | 4.3877 | **57.3395** | **0.2102** | **14680.40** |
| **BPE**                   | ASCII  |     0.000095 |     1.2858 | 5.0952 |     57.8477 |     0.2149 |     14922.79 |
| **WordPiece**             | Bangla |     0.000056 |     1.2896 | 4.3363 |     58.0200 |     0.2177 |     15091.11 |
| **WordPiece**             | ASCII  |     0.000057 |     1.3031 | 5.0277 |     58.6248 |     0.2225 |     15391.12 |
| **SentencePiece Unigram** | Bangla | **0.000000** |     1.3097 | 4.2699 |     58.9222 |     0.2310 |     15773.49 |
| **SentencePiece Unigram** | ASCII  | **0.000000** |     1.3736 | 4.7695 |     61.7988 |     0.2540 |     17306.75 |

### Key observations

* **BPE** achieved the best overall intrinsic performance with the **lowest fertility, WFR, average tokens, and variance**.
* **WordPiece** ranked second and consistently outperformed SentencePiece Unigram.
* **SentencePiece Unigram** achieved **zero OOV** but produced the highest fragmentation and variance.
* Across all three tokenizers, the **native Bangla script consistently outperformed ASCII transliteration** in terms of segmentation compactness and stability.
* **ASCII transliteration** naturally resulted in higher **Characters per Token (CPT)** because transliterated words are represented using longer Latin character sequences. This increase reflects the script representation rather than improved tokenization quality.
